### PR TITLE
Implement support for Debug.Trace.trace

### PR DESCRIPTION
**Pull request by bollu**
- [x] Add a test case
- [x] Implement `debugBelch2`
This PR depends on #133, so the history is understandably messy

### CODE REVIEW

**Comment by bollu** [comment]
@TerrorJack It compiles! and prints the error :)

File: asterius/rts/rts.messages.mjs [hunk 1]
```diff
@@ -0,0 +1,11 @@
+export class Messages {
+    constructor(memory) {
+        this.memory = memory;
+        Object.seal(this);
+    }
+
+    debugBelch2(fmt, arg) {
```
Comment (TerrorJack) [inline asterius/rts/rts.messages.mjs]: iirc all `debugBelch2` calls are `debugBelch2("%s", c_str)`, so it makes sense to simply ignore the first format string here.

File: asterius/rts/rts.memory.mjs [hunk 2]
```diff
@@ -56,6 +56,19 @@ export class Memory {
   i64LoadU16(p) { return BigInt(this.dataView.getUint16(Memory.unTag(p), true)); }
   heapAlloced(p) { return Memory.unTag(p) >= (this.staticMBlocks << Math.log2(rtsConstants.mblock_size)); }
   strlen(_str) { return this.i8View.subarray(Memory.unTag(_str)).indexOf(0); }
+  strLoad(_str) {
+      let p = Memory.unTag(_str);
+      let s = "";
+      let i = 0;
+      // NOTE: passing a non-null terminated string will give you garbage,
+      // as you deserve.
+      while(1) {
```
Comment (TerrorJack) [inline asterius/rts/rts.memory.mjs]: Can we use a `for` loop here? We usually use `while(true)` only when the loop range is not trivial to calculate (e.g. walking a stack and stopping for specific frame types)

File: asterius/rts/rts.messages.mjs [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+        console.error(this.memory.strLoad(fmt), this.memory.strLoad(arg));
```
Comment (TerrorJack) [inline asterius/rts/rts.messages.mjs]: `console.error` logs to the `node` process's standard error, keep in mind that we have our own `stderr` and that custom `stderr` buffer is checked by `ghc-testsuite`. So do make sure we log to our own `stderr` as well.

File: asterius/rts/rts.messages.mjs (hunk 1 above)
Comment (bollu) [reply asterius/rts/rts.messages.mjs]: True, but it doesn't _cost_ us to implement a general formatter, and we're correct to this spec, so I don't see a problem per se?

File: asterius/rts/rts.memory.mjs (hunk 2 above)
Comment (bollu) [reply asterius/rts/rts.memory.mjs]: Done.

File: asterius/rts/rts.messages.mjs (hunk 3 above)
Comment (bollu) [reply asterius/rts/rts.messages.mjs]: Done.

File: asterius/rts/rts.memory.mjs [hunk 4]
```diff
@@ -56,6 +56,16 @@ export class Memory {
   i64LoadU16(p) { return BigInt(this.dataView.getUint16(Memory.unTag(p), true)); }
   heapAlloced(p) { return Memory.unTag(p) >= (this.staticMBlocks << Math.log2(rtsConstants.mblock_size)); }
   strlen(_str) { return this.i8View.subarray(Memory.unTag(_str)).indexOf(0); }
+  strLoad(_str) {
+      let p = Memory.unTag(_str);
+      let s = "";
+
+      for(let i = 0; i < this.strlen(_str); ++i) {
```
Comment (TerrorJack) [inline asterius/rts/rts.memory.mjs]: This loop traverses the c string twice, first finding its length and then converting to js string. We don't need two traversals.

Comment (bollu) [reply asterius/rts/rts.memory.mjs]: Reverting back to old `while(1) {}` style

**Pull request merged**